core/src/consensus/stake/actions.rs

```diff
@@ -247,16 +347,283 @@ mod tests {
+        }
+        let result = double_vote_verification_result(
+            ConsensusMessageInfo {
+                height: 2,
+                view: 0,
+                step: Step::Precommit,
+                block_hash: hash1,
+                signer_index: 1,
+            },
+            ConsensusMessageInfo {
+                height: 2,
+                view: 0,
+                step: Step::Precommit,
+                block_hash: hash2,
+                signer_index: 0,
+            },
+            &|v| v,
+            &|v| v,
+        );
+        match result {
+            Err(VerifyError::Syntax(SyntaxError::InvalidCustomAction(ref s)))
+                if s.contains("Two messages are signed by different signers") => {}
+            _ => panic!(),
+        }
+    }
+
+    #[test]
+    fn double_vote_verify_strange_sig1() {
+        let vote_step_twister = |original: VoteStep| VoteStep {
+            height: original.height + 1,
+            view: original.height + 1,
+            step: original.step,
+        };
+        let result = double_vote_verification_result(
+            ConsensusMessageInfo {
+                height: 2,
+                view: 0,
+                step: Step::Precommit,
+                block_hash: None,
+                signer_index: 0,
+            },
+            ConsensusMessageInfo {
+                height: 2,
+                view: 0,
+                step: Step::Precommit,
+                block_hash: Some(H256::random()),
+                signer_index: 0,
+            },
+            &vote_step_twister,
+            &|v| v,
+        );
+        let expected_err = Err(VerifyError::Syntax(SyntaxError::InvalidCustomAction(String::from(
+            "Schnorr signature verification fails",
+        ))));
+        assert_eq!(result, expected_err);
+    }
+
+    #[test]
+    fn double_vote_verify_strange_sig2() {
+        let block_hash_twister = |original: Option<H256>| {
+            original.map(|hash| {
+                let mut twisted = H256::random();
+                while twisted == hash {
+                    twisted = H256::random();
+                }
+                twisted
+            })
+        };
+        let result = double_vote_verification_result(
+            ConsensusMessageInfo {
+                height: 2,
+                view: 0,
+                step: Step::Precommit,
+                block_hash: None,
+                signer_index: 0,
+            },
+            ConsensusMessageInfo {
+                height: 2,
+                view: 0,
+                step: Step::Precommit,
+                block_hash: Some(H256::random()),
+                signer_index: 0,
+            },
+            &|v| v,
+            &block_hash_twister,
+        );
+        let expected_err = Err(VerifyError::Syntax(SyntaxError::InvalidCustomAction(String::from(
+            "Schnorr signature verification fails",
+        ))));
+        assert_eq!(result, expected_err);
+    }
```

Comment:
Would you like to add a test that a message is from an older block?

Reply:
I didn't catch the meaning of "a message from an older block" .
In this test I manually set the block height of the messages. So any block can be tested.

Reply:
I was referring that you should catch this kind of [bug](
https://github.com/CodeChain-io/codechain/pull/1640#discussion_r306205318) too. I didn't know that you are just test 'verify' function. Let it be handled in e2e test then.